### PR TITLE
KIN 216 - Category and Tags tables and seeders

### PR DESCRIPTION
**Pull request by CarlosJCD**
[KIN-216](https://loopcrack.atlassian.net/browse/KIN-216) Medium [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=loopcrack-org&utm_campaign=badge_info)

### Problem

Store the records of the categories and tags related to the categories. 

### Solution

Create a table for the categories and its tags on the database

## Acceptance Criteria

- It must store the following data: 
    - Category name
    - Category icon path
    - Category image path
- It must store the tags in a table with the following data:
    - Tags name
    - Id of the category associated

### Testing

- [x] Via php spark and with the seeders, run the Category Seeder to generate 2 records of a category.
- [x] Via php spark and with the Category Seeder previously runed, run the TagsSeeder to generate 4 records of tags

## Resources

- [Database Design Document](https://loopcrack.atlassian.net/wiki/spaces/Kinub/pages/1277953/Database+Design)

## Demo

[KIN-216]: https://loopcrack.atlassian.net/browse/KIN-216?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by Oswaldo-Chan — APPROVED** [review]

**Review by CheloChidoone — COMMENTED** [review]
@CarlosJCD The `idCategory` field should be able to be `NULL` as well right?, because not all tags are going to belong to a category, some are going to just be for a product, have you thought of how that is going to work?

**Comment by CarlosJCD** [comment]
> @CarlosJCD The `idCategory` field should be able to be `NULL` as well right?, because not all tags are going to belong to a category, some are going to just be for a product, have you thought of how that is going to work?

@CheloChidoone actually we are only storing the tags related to the categories because the product tags are specific and doesn't have a functionality beyond of being informative about the product. 

However I think it might be a better practice to change the model and database table names to category_tags so that we can avoid this confusion in the future.

**Review by CheloChidoone — APPROVED** [review]

**Review by Cristianpan — APPROVED** [review]

File: backup/Database.sql [hunk 1]
```diff
@@ -29,7 +54,7 @@ CREATE TABLE `Email` (
   PRIMARY KEY (`id`),
   KEY `Correos_FK` (`idTypeEmail`),
   CONSTRAINT `Correos_FK` FOREIGN KEY (`idTypeEmail`) REFERENCES `type_email` (`id`)
-) ENGINE=InnoDB AUTO_INCREMENT=5 DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci;
+) ENGINE=InnoDB AUTO_INCREMENT=6 DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci;
```
Comment (LUGAMAFE) [inline backup/Database.sql:57]: here too

File: backup/Database.sql [hunk 2]
```diff
@@ -41,6 +66,32 @@ LOCK TABLES `Email` WRITE;
...
+/*!40101 SET @saved_cs_client     = @@character_set_client */;
+/*!50503 SET character_set_client = utf8mb4 */;
+CREATE TABLE `tags` (
+  `id` int NOT NULL AUTO_INCREMENT,
+  `name` varchar(60) COLLATE utf8mb4_unicode_ci NOT NULL,
+  `idCategory` int NOT NULL,
+  PRIMARY KEY (`id`),
+  KEY `Tags_FK` (`idCategory`),
+  CONSTRAINT `Tags_FK` FOREIGN KEY (`idCategory`) REFERENCES `categories` (`id`)
+) ENGINE=InnoDB AUTO_INCREMENT=21 DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci;
```
Comment (LUGAMAFE) [inline backup/Database.sql:83]: here too

File: app/Database/Seeds/CategorySeeder.php [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+use CodeIgniter\Database\Seeder;
+
+class CategorySeeder extends Seeder
+{
+    public function run()
+    {
+        $data = [
+            [
+                "name" => "Software",
+                "image" => md5(uniqid(rand(), true)),
```
[the remark this answers is not in the document]
Comment (CarlosJCD) [reply app/Database/Seeds/CategorySeeder.php:13]: El nombre de la imagen de la categoría. Se realiza el proceso de `md5(uniqid(rand(),true))` con la finalidad de que se almacenen las imágenes con hash únicos como nombres

[the remark this answers is not in the document]
Comment (LUGAMAFE) [reply app/Database/Seeds/CategorySeeder.php:13]: @CarlosJCD pero en la imagen no deberiamos estar almacenando la url de la imagen? te refieres que el nombre de las imagenes seran randoms? y si es asi de que nos sirve que pongas aqui en el seeder este valor? `md5(uniqid(rand(), true))` si no hay ninguna imagen guardada por el momento con ese hash?

[the remark this answers is not in the document]
Comment (CarlosJCD) [reply app/Database/Seeds/CategorySeeder.php:13]: Teniendo en cuenta las rutas para las imagenes podriamos ampliar el campo de varchar, sin embargo, la razon por la que en el seeder almacenamos ese hash aunque no haya imagenes es para corroborar que en efecto la tabla si llegue a almacenar el valor, ya que ese tipo de valores se estaran usando con las imagenes.

**Review by LUGAMAFE — APPROVED** [review]

**Pull request merged**